feat: 添加区块双向绑定属性选择功能到事件表单

0x7A7A6572 opened the pull request:
在 BlockEventForm.vue 中添加了双向绑定属性选择器，允许用户通过悬浮图标查看并选择可用属性来自动生成v-model事件。

English | [简体中文](https://github.com/opentiny/tiny-engine/blob/develop/.github/PULL_REQUEST_TEMPLATE/PULL_REQUEST_TEMPLATE.zh-CN.md)

# PR

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our [Commit Message Guidelines](https://github.com/opentiny/tiny-engine/blob/develop/CONTRIBUTING.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] Built its own designer, fully self-validated

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Background and solution
<!--
1. Describe the problem and the scenario.
2. New features need to be described and attached with renderings.
3. Screenshots or GIFs involving UI/Interaction changes/Bugfix before and after modification are required.
-->

### What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

### What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

添加区块-事件管理中双向绑定属性的事件，对应schema  events  的 onUpdate:eventName
![Snipaste_2025-05-30_03-12-59](https://github.com/user-attachments/assets/4eadb42e-5f86-4e35-a670-6d57b9098105)
![Snipaste_2025-05-30_03-13-39](https://github.com/user-attachments/assets/ed306fca-99eb-4c9d-9b1d-daf825e5b627)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added a popover to the event name input, allowing users to quickly bind update events to block properties.
  * Users can select properties from a list to auto-generate "onUpdate:" events and update labels accordingly.

* **Improvements**
  * Enhanced event name validation to recognize and support auto-generated update events.
  * Improved UI feedback and styling for event binding and selection.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

0x7A7A6572 committed:
feat: 添加区块双向绑定属性选择功能到事件表单
在 BlockEventForm.vue 中添加了双向绑定属性选择器，允许用户通过悬浮图标查看并选择可用属性来自动生成v-model事件。

chilingling commented:
![截屏2025-06-10 19 25 15](https://github.com/user-attachments/assets/b64f1b03-b575-4caf-a801-40b42428d86c)
屏幕宽度窄的时候，会被遮挡，可以考虑下 suffix 插槽实现，或者是增加白色背景

chilingling commented on packages/plugins/block/src/BlockEventForm.vue:
```diff
@@ -1,7 +1,31 @@
   <tiny-form label-position="left" label-width="60px" show-message :model="formData" :rules="rules">
     <tiny-form-item label="事件名" prop="eventName">
-      <tiny-input v-model="formData.eventName" :placeholder="eventNameTip" @blur="changeEventName"></tiny-input>
+      <tiny-input v-model="formData.eventName" :placeholder="eventNameTip" @blur="changeEventName">
+        <template #suffix>
+          <tiny-popover v-model="state.showPopover" placement="bottom-end" trigger="hover" popperClass="option-popper">
+            <template #reference>
+              <tiny-icon-rich-text-link :class="{ 'bind-propertys': isUpdateEvent }" />
+            </template>
+            <div class="property-list">
+              <div class="property-list-title">
+                <tiny-icon-rich-text-link />
+                双向绑定属性
+              </div>
+              <ul>
+                <li
+                  v-for="(item, index) in state.propertys"
+                  :key="index"
+                  :class="{ existed: state.events.hasOwnProperty('onUpdate:' + item.property) }"
+                  @click="usePropertysToBeEvent(item)"
+                >
+                  <div>{{ item.property }}</div>
+                </li>
+              </ul>
+            </div>
+          </tiny-popover>
+        </template>
+      </tiny-input>
```
@0x7A7A6572 这个 AI 的建议可以改一下，使用 `Object.prototype.hasOwnProperty.call` 避免其他安全问题

chilingling replied:
或者可以这样子：
1. 使用 computed 计算事件名：

```javascript
const nameList = computed(() => new Set(Object.keys(state.events)))
```

2. 在模板中使用 `has` 进行判断

```vue
:class = "{ existed: nameList.has(`onUpdate:${item.property}`) }"
```

0x7A7A6572 committed:
style: optimized the suffix style of the time input box

0x7A7A6572 committed:
fix: fix the unsafe hasOwnProperty and optimize popover popup closure

- 将 popover 触发方式从 hover 改为 manual，让showPopover正确响应
- 修复不安全的 hasOwnProperty

chilingling commented:
![image](https://github.com/user-attachments/assets/16208007-0869-479a-9ebb-15bd457eaa62)

需要监听外部点击事件，在恰当的时机关闭该 popup

0x7A7A6572 committed:
feat: add block collapsed state sharing

chilingling commented:
1. 建议弹窗收起的监听不仅仅是监听事件设置项的收起。可以尝试监听全局的点击收起，或者是使用 TinyPopover 的 click，会自动收起。

问题示例1：点击之后不收起设置设置弹窗，点击具体的属性设置，弹窗位置显示在了属性设置的面板。（属性设置面板被拉长了）
![image](https://github.com/user-attachments/assets/3bf29bdd-2cca-4fb3-ab91-4b8f02fe027f)

问题示例2:
删除当前编辑事件，弹窗不收起
![截屏2025-06-19 14 30 09](https://github.com/user-attachments/assets/d6fbf766-73ab-4805-89f9-e1255f34965e)

0x7A7A6572 commented:
之前有尝试过，TinyPopover  的 click 也解决不了这问题，监听面板全局的收起变化可以，或者用回TinyPopover  的 hover（确认不会自动关闭），我先按照全局收起变化的方案修改

0x7A7A6572 commented:
全局收起还不够，如  问题示例1 ，或者是其他的会更改布局的操作，无法得到好的解决，再多加事件监听反而偏离主题了，有什么好的建议吗，舍弃一部分的操作体验使用 hover 可能好一些？

0x7A7A6572 committed:
fix: 移除收起监听，简化事件表单的popover交互方式，改为hover触发

0x7A7A6572 committed:
fix: 修复悬浮框显示状态控制问题

添加showPopover状态控制悬浮框的显示，通过对 popover内容显隐使 popover 失去焦点，以达成关闭动作

0x7A7A6572 committed:
fix: 修复事件名验证逻辑并优化属性与事件管理

重构事件表单组件，将属性和事件从state中提取为独立计算属性
修复事件名验证逻辑，支持onUpdate:前缀并添加属性存在性检查
更新事件名提示信息，明确命名规则

chilingling commented:
<img width="533" height="383" alt="截屏2025-07-22 14 51 04" src="https://github.com/user-attachments/assets/2b767c35-de56-4b6f-a9d3-6bbd50023283" />

@0x7A7A6572  建议增加一下最大高度，超过最大高度之后通过滚动来选择

其他就没啥问题啦。🎉

chilingling commented on packages/plugins/block/src/BlockEventForm.vue:
```diff
@@ -55,8 +94,35 @@ export default {
+              const propertyMatched = propertys.value.some((item) => item.property === value.replace('onUpdate:', ''))
+              return matched && propertyMatched
+                ? callback()
+                : callback(new Error('无效的 onUpdate 事件名, onUpdate 事件需在属性设置中定义'))
```
这里提示应该是需在属性设置中定义对应的属性？或者说 `onUpdate:propertyName` 需要有对应的 `propertyName` 在属性中定义？

0x7A7A6572 committed:
style: 优化属性绑定列表滑动以及校验文本提示
